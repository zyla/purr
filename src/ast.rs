use crate::symbol::Symbol;

#[derive(Debug)]
pub struct SourceSpan {
    pub start: usize,
    pub end: usize,
}

#[derive(Debug)]
pub struct Located<T>(pub SourceSpan, pub T);

#[derive(Debug)]
pub struct Comment(pub String);

#[derive(Debug)]
pub struct Commented<T>(pub Vec<Comment>, pub T);

pub type Module = Located<Commented<ModuleInner>>;

#[derive(Debug)]
pub struct ModuleInner {
    pub name: Symbol,
    pub exports: Option<Vec<DeclarationRef>>,
    pub declarations: Vec<Declaration>,
}

#[derive(Debug)]
pub enum DeclarationRef {
    // TODO
}

pub type Declaration = Located<Commented<DeclarationKind>>;

#[derive(Debug)]
pub enum DeclarationKind {
    ValueDeclaration(ValueDeclaration),
}

#[derive(Debug)]
pub struct ValueDeclaration {
    pub ident: Symbol,

    // TODO: what is this?
    // pub name: NameKind

    // TODO
    // pub binders: Vec<Binder>
    pub expr: Vec<GuardedExpr>,
}

#[derive(Debug)]
pub struct GuardedExpr {
    pub guards: Vec<Guard>,
    pub expr: Expr,
}

#[derive(Debug)]
pub enum Guard {
    // TODO
}

pub type Expr = Located<ExprKind>;

#[derive(Debug)]
pub enum ExprKind {
    Literal(Literal<Expr>),
    // TODO
}

#[derive(Debug)]
pub enum Literal<T> {
    Integer(i32),
    Float(f64),
    String(String),
    Char(char),
    Boolean(bool),
    Array(Vec<T>),
    Object(Vec<(Symbol, T)>),
}
