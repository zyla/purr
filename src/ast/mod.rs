pub mod declarations;
pub mod expr;
pub mod meta;
pub mod types;

use crate::ModuleId;
pub use declarations::*;
pub use expr::*;
pub use meta::*;
pub use types::*;

use crate::symbol::Symbol;

#[salsa::interned]
pub struct AbsoluteName {
    pub module: ModuleId,
    pub name: Symbol,
}

#[salsa::interned]
pub struct QualifiedName {
    pub module: Option<ModuleId>,
    pub name: Symbol,
}

impl QualifiedName {
    pub fn is_actually_qualified(&self, db: &dyn crate::Db) -> bool {
        self.module(db).is_some()
    }

    pub fn new_unqualified(db: &dyn crate::Db, name: Symbol) -> Self {
        Self::new(db, None, name)
    }

    pub fn new_qualified(db: &dyn crate::Db, module: ModuleId, name: Symbol) -> Self {
        Self::new(db, Some(module), name)
    }
}
