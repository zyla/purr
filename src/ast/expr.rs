use super::{Declaration, Located, Type};
use crate::ast::QualifiedName;
use crate::string::PSChar;
use crate::string::PSString;
use crate::symbol::Symbol;
use ordered_float::OrderedFloat;

pub type Expr = Located<ExprKind>;

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub enum ExprKind {
    Literal(Literal<Expr>),

    /// Infix operator sequence with unknown precedence
    Infix(Box<Expr>, Vec<(InfixOp, Expr)>),

    /// Record field accessor
    Accessor(Box<Expr>, Symbol),

    RecordUpdate(Box<Expr>, RecordUpdate),

    Var(QualifiedName),

    /// Standalone operator
    Operator(InfixOp),

    DataConstructor(QualifiedName),

    App(Box<Expr>, Vec<Expr>),

    Lam(Vec<Pat>, Box<Expr>),

    Case {
        exprs: Vec<Expr>,
        branches: Vec<CaseBranch>,
    },

    If {
        cond: Box<Expr>,
        then_: Box<Expr>,
        else_: Box<Expr>,
    },

    Typed(Box<Expr>, Box<Type>),

    Let {
        decls: Vec<Declaration>,
        body: Box<Expr>,
    },

    Wildcard,

    // Pseudo-expression, used only as an intermediate value during parsing.
    RecordUpdateSuffix(RecordUpdate),

    // Pseudo-expression, used only as an intermediate value during parsing.
    NamedPat(Symbol, Box<Expr>),

    Do(Vec<DoItem>),

    Ado(Vec<DoItem>, Box<Expr>),

    Negate(Box<Expr>),
}

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub enum InfixOp {
    Symbol(QualifiedName),
    Backtick(Box<Expr>),
}

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub enum DoItem {
    Let(Vec<Declaration>),
    Expr(Expr),
    Bind(Pat, Expr),
}

type RecordUpdate = Vec<(Symbol, Expr)>;

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub enum RecordLiteralOrUpdate {
    Literal(Vec<(Symbol, Expr)>),
    Update(Vec<(Symbol, Expr)>),
}

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub struct CaseBranch {
    pub pats: Vec<Pat>,
    pub expr: PossiblyGuardedExpr,
}

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub enum PossiblyGuardedExpr {
    Unconditional(Expr),
    Guarded(Vec<GuardedExpr>),
}

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub struct GuardedExpr {
    pub guards: Vec<Guard>,
    pub expr: Expr,
}

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub enum Guard {
    Expr(Expr),
    Bind(Pat, Expr),
}

pub type Pat = Located<PatKind>;

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub enum PatKind {
    Literal(Literal<Pat>),

    /// Infix operator sequence with unknown precedence
    Infix(Box<Pat>, Vec<(QualifiedName, Pat)>),

    Var(Symbol),

    DataConstructorApp(QualifiedName, Vec<Pat>),

    Wildcard,

    Named(Symbol, Box<Pat>),

    Typed(Box<Pat>, Box<Type>),
}

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub enum Literal<T> {
    Integer(i64),
    Float(OrderedFloat<f64>),
    String(PSString),
    Char(PSChar),
    Boolean(bool),
    Array(Vec<T>),
    Object(Vec<(Symbol, T)>),
}

#[test]
fn test_size() {
    // Note: it was 56 before adding multi-case.
    assert_eq!(std::mem::size_of::<Expr>(), 72);
}
