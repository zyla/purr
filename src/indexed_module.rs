use crate::ast::AbsoluteName;
use crate::ast::CaseBranch;
use crate::ast::DataConstructorDeclaration;
use crate::ast::DataDeclType;
use crate::ast::Declaration;
use crate::ast::DeclarationKind;
use crate::ast::Fundep;
use crate::ast::Located;
use salsa::DebugWithDb;
use std::iter::Peekable;
use std::path::PathBuf;

use crate::ast::Type;
use crate::ast::TypeDeclarationData;
use crate::ast::TypeParameter;
use crate::ParsedModule;
use crate::{Db, ModuleId};
use crate::{Diagnostic, Diagnostics};
use fxhash::FxHashMap;
use std::collections::hash_map::Entry;

#[derive(PartialEq, Eq, Clone, Debug, DebugWithDb)]
pub enum TypeDecl {
    Data(DataDecl),
    Type(TypeSynonymDecl),
    TypeClass(TypeClassDecl),
}

#[derive(PartialEq, Eq, Clone, Debug, DebugWithDb)]
pub struct DataDecl {
    pub type_: DataDeclType,
    pub name: AbsoluteName,
    pub params: Vec<TypeParameter>,
    pub kind: Option<Type>,
    pub constructors: Vec<DataConstructorDeclaration>,
}

#[derive(PartialEq, Eq, Clone, Debug, DebugWithDb)]
pub struct TypeSynonymDecl {
    pub name: AbsoluteName,
    pub params: Vec<TypeParameter>,
    pub body: Type,
}

#[derive(PartialEq, Eq, Clone, Debug, DebugWithDb)]
pub struct TypeClassDecl {
    pub name: AbsoluteName,
    pub constraints: Vec<Type>,
    pub params: Vec<TypeParameter>,
    pub fundeps: Vec<Fundep>,
    pub methods: Vec<TypeDeclarationData>,
}

#[derive(PartialEq, Eq, Clone, Debug, DebugWithDb)]
pub struct ValueDecl {
    pub name: AbsoluteName,
    pub type_: Option<Type>,

    /// Empty equations means foreign import.
    /// TODO: maybe make it more explicit?
    pub equations: Vec<CaseBranch>,
}

struct ModuleIndexer<'a> {
    db: &'a dyn Db,
    // Note: Using `FxHashMap` mostly because we want deterministic order for snapshots.
    types: FxHashMap<AbsoluteName, TypeDecl>,
    values: FxHashMap<AbsoluteName, ValueDecl>,
    classes: FxHashMap<AbsoluteName, TypeClassDecl>,
    module_id: ModuleId,
    filename: PathBuf,
    // TODO: Add map with Decl to SourceSpan
}

impl<'a> ModuleIndexer<'a> {
    fn index(&mut self, module: ParsedModule) {
        let db = self.db;
        let mut iter = module.ast.declarations.iter().peekable();
        while let Some(src_decl) = iter.peek().copied() {
            use DeclarationKind::*;

            let ref_loc = src_decl.0.start;

            match &***src_decl {
                Data {
                    type_,
                    name,
                    params,
                    kind,
                    constructors,
                } => {
                    let abs_name = AbsoluteName::new(db, self.module_id, *name);
                    let mut params = params.clone();
                    to_relative_params(abs_name, ref_loc, &mut params);

                    let mut constructors = constructors.clone();
                    constructors
                        .iter_mut()
                        .for_each(|c| c.to_relative_span(abs_name, ref_loc));

                    match self.types.entry(abs_name) {
                        Entry::Occupied(_) => {
                            Diagnostics::push(
                                db,
                                Diagnostic::new(
                                    src_decl.span().start,
                                    src_decl.span().end,
                                    format!("Duplicate type declaration {}", name.text(db)),
                                    module.filename.to_string_lossy().into(),
                                ),
                            );
                        }
                        Entry::Vacant(e) => {
                            e.insert(TypeDecl::Data(DataDecl {
                                type_: *type_,
                                name: abs_name,
                                params,
                                kind: kind.clone(),
                                constructors,
                            }));
                        }
                    }
                    iter.next();
                }

                TypeSynonym { name, params, body } => {
                    let abs_name = AbsoluteName::new(db, self.module_id, *name);

                    match self.types.entry(abs_name) {
                        Entry::Occupied(_) => {
                            Diagnostics::push(
                                db,
                                Diagnostic::new(
                                    src_decl.span().start,
                                    src_decl.span().end,
                                    format!("Duplicate type declaration {}", name.text(db)),
                                    module.filename.to_string_lossy().into(),
                                ),
                            );
                        }
                        Entry::Vacant(e) => {
                            let mut params = params.clone();
                            to_relative_params(abs_name, ref_loc, &mut params);

                            let mut body = body.clone();
                            body.to_relative_span(abs_name, ref_loc);

                            e.insert(TypeDecl::Type(TypeSynonymDecl {
                                name: abs_name,
                                params,
                                body,
                            }));
                        }
                    }
                    iter.next();
                }

                KindSignature { .. } => todo!(),

                Role(_) => todo!(),

                TypeSignature(sig) => {
                    iter.next();
                    match iter.peek() {
                        Some(x) if matches!(&****x, ValueDeclaration(_)) => {
                            self.parse_value_decl(
                                &mut iter,
                                Some(Located::new(src_decl.span(), sig.clone())),
                                ref_loc,
                            );
                        }
                        _ => {
                            Diagnostics::push(
                                db,
                                Diagnostic::new(
                                    src_decl.span().start,
                                    src_decl.span().end,
                                    format!(
                                        "Type signature of {} should be followed by its definition",
                                        sig.ident.text(db)
                                    ),
                                    module.filename.to_string_lossy().into(),
                                ),
                            );
                        }
                    }
                }
                ValueDeclaration(_) => {
                    self.parse_value_decl(&mut iter, None, ref_loc);
                }
                Destructuring { .. } => {
                    Diagnostics::push(
                        db,
                        Diagnostic::new(
                            src_decl.span().start,
                            src_decl.span().end,
                            "Invalid top-level destructuring".to_string(),
                            module.filename.to_string_lossy().into(),
                        ),
                    );
                    iter.next();
                }
                ForeignValue { name, type_ } => {
                    let abs_name = AbsoluteName::new(db, self.module_id, *name);

                    match self.values.entry(abs_name) {
                        Entry::Occupied(_) => {
                            Diagnostics::push(
                                db,
                                Diagnostic::new(
                                    src_decl.span().start,
                                    src_decl.span().end,
                                    format!("Duplicate value declaration {}", name.text(db)),
                                    self.filename.to_string_lossy().into(),
                                ),
                            );
                        }
                        Entry::Vacant(e) => {
                            let mut type_ = type_.clone();
                            type_.to_relative_span(abs_name, ref_loc);

                            e.insert(ValueDecl {
                                name: abs_name,
                                type_: Some(type_),
                                equations: vec![],
                            });
                        }
                    }
                    iter.next();
                }
                Class(type_class_decl) => {
                    let abs_name = AbsoluteName::new(db, self.module_id, type_class_decl.name);

                    match self.classes.entry(abs_name) {
                        Entry::Occupied(_) => {
                            Diagnostics::push(
                                db,
                                Diagnostic::new(
                                    src_decl.span().start,
                                    src_decl.span().end,
                                    format!(
                                        "Duplicate typeclass declaration {}",
                                        type_class_decl.name.text(db)
                                    ),
                                    self.filename.to_string_lossy().into(),
                                ),
                            );
                        }
                        Entry::Vacant(e) => {
                            let mut constraints = type_class_decl.constraints.clone();
                            constraints
                                .iter_mut()
                                .for_each(|c| c.to_relative_span(abs_name, ref_loc));

                            let mut params = type_class_decl.params.clone();
                            to_relative_params(abs_name, ref_loc, &mut params);

                            let mut methods = type_class_decl.methods.clone();
                            methods
                                .iter_mut()
                                .for_each(|decl| decl.r#type.to_relative_span(abs_name, ref_loc));

                            e.insert(TypeClassDecl {
                                name: abs_name,
                                constraints,
                                params,
                                fundeps: type_class_decl.fundeps.clone(),
                                methods,
                            });
                        }
                    }

                    iter.next();
                }
                InstanceChain(_) => todo!(),
                Operator { .. } => todo!(),
            }
        }
    }

    fn parse_value_decl<'m>(
        &mut self,
        iter: &mut Peekable<impl Iterator<Item = &'m Declaration>>,
        sig: Option<Located<TypeDeclarationData>>,
        reference_loc: usize,
    ) {
        let db = self.db;
        let Some(first_decl) = iter.next() else {
            panic!("parse_value_decl should be called when next is ValueDeclaration")
        };
        let DeclarationKind::ValueDeclaration(first) = &***first_decl else {
            panic!("parse_value_decl should be called when next is ValueDeclaration")
        };

        let abs_name = AbsoluteName::new(db, self.module_id, first.ident);

        let ty = match sig {
            None => None,
            Some(sig) => {
                if sig.ident == first.ident {
                    let mut sig_ty = sig.into_inner().r#type;
                    sig_ty.to_relative_span(abs_name, reference_loc);
                    Some(sig_ty)
                } else {
                    Diagnostics::push(
                        db,
                        Diagnostic::new(
                            sig.span().start,
                            sig.span().end,
                            format!(
                                "Type signature of {} should be followed by its definition",
                                sig.ident.text(db)
                            ),
                            self.filename.to_string_lossy().into(),
                        ),
                    );
                    None
                }
            }
        };
        let mut equations = vec![first.clone().into()];

        //TODO: Make spans relative in equations SourceSpan
        while let Some(src_decl) = iter.peek() {
            let DeclarationKind::ValueDeclaration(decl) = &****src_decl else {
                break;
            };
            if decl.ident != first.ident {
                // Start of another chain of equations
                break;
            }
            equations.push(decl.clone().into());
            iter.next();
        }

        match self.values.entry(abs_name) {
            Entry::Occupied(_) => {
                Diagnostics::push(
                    db,
                    Diagnostic::new(
                        first_decl.span().start,
                        first_decl.span().end,
                        format!("Duplicate value declaration {}", first.ident.text(db)),
                        self.filename.to_string_lossy().into(),
                    ),
                );
            }
            Entry::Vacant(e) => {
                e.insert(ValueDecl {
                    name: abs_name,
                    type_: ty,
                    equations,
                });
            }
        }
    }
}

fn to_relative_params(
    abs_name: AbsoluteName,
    reference_loc: usize,
    params: &mut [(crate::symbol::Symbol, Option<Type>)],
) {
    params.iter_mut().for_each(|(_, type_opt)| {
        if let Some(typ) = type_opt {
            typ.to_relative_span(abs_name, reference_loc)
        }
    })
}

#[derive(PartialEq, Eq, Clone, Debug, DebugWithDb)]
pub struct IndexedModule {
    pub module_id: ModuleId,
    pub types: FxHashMap<AbsoluteName, TypeDecl>,
    pub values: FxHashMap<AbsoluteName, ValueDecl>,
    pub classes: FxHashMap<AbsoluteName, TypeClassDecl>,
}

#[salsa::tracked]
pub fn indexed_module(db: &dyn Db, module_id: ModuleId) -> IndexedModule {
    let module = crate::parsed_module(db, module_id);
    let mut indexer = ModuleIndexer {
        db,
        types: Default::default(),
        values: Default::default(),
        classes: Default::default(),
        module_id,
        filename: module.filename.clone(),
    };
    indexer.index(module);
    IndexedModule {
        module_id,
        types: indexer.types,
        values: indexer.values,
        classes: indexer.classes,
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::utils::tests::*;
    use indoc::indoc;
    use insta::{self, assert_snapshot};

    fn index_module(input: &str) -> String {
        let db = &mut crate::Database::test_single_file_db(input);
        let module_id = parse_module_id(input, db);
        format!(
            "{:#?}",
            (
                indexed_module(db, module_id).into_debug_all(db),
                indexed_module::accumulated::<Diagnostics>(db, module_id)
            )
        )
        .drop_salsa_id()
    }

    #[test]
    fn index_data_decl() {
        assert_snapshot!(index_module(indoc!(
            "
        module Test where
        data Foo = Foo | Bar
        "
        )));
    }

    #[test]
    fn duplicate_data_decl() {
        assert_snapshot!(index_module(indoc!(
            "
        module Test where
        data Foo
        data Foo
        "
        )));
    }

    #[test]
    fn type_sig_without_equations_1() {
        assert_snapshot!(index_module(indoc!(
            "
        module Test where
        foo :: Int
        "
        )));
    }

    #[test]
    fn type_sig_without_equations_2() {
        assert_snapshot!(index_module(indoc!(
            "
        module Test where
        foo :: Int
        bar = 1
        "
        )));
    }

    #[test]
    fn type_sig_without_equations_3() {
        assert_snapshot!(index_module(indoc!(
            "
        module Test where
        foo :: Int
        data Foo
        "
        )));
    }

    #[test]
    fn value_decl() {
        assert_snapshot!(index_module(indoc!(
            "
        module Test where
        foo = 1
        "
        )));
    }

    #[test]
    fn value_decl_with_signature() {
        assert_snapshot!(index_module(indoc!(
            "
        module Test where
        foo :: Int
        foo = 1
        "
        )));
    }

    #[test]
    fn value_decl_many_equations() {
        assert_snapshot!(index_module(indoc!(
            "
        module Test where
        foo 1 = 1
        foo 2 = 2
        "
        )));
    }

    #[test]
    fn type_alias() {
        assert_snapshot!(index_module(indoc!(
            "
        module Test where
        type Foo = Int
        "
        )));
    }

    #[test]
    fn type_and_data_decl_conflict() {
        assert_snapshot!(index_module(indoc!(
            "
        module Test where
        type Foo = Int
        data Foo
        "
        )));
    }

    #[test]
    fn foreign_import() {
        assert_snapshot!(index_module(indoc!(
            "
        module Test where
        foreign import foo :: Int
        "
        )));
    }

    #[test]
    fn duplicate_value() {
        assert_snapshot!(index_module(indoc!(
            "
        module Test where
        foo = 1
        bar = 2
        foo = 1
        "
        )));
    }

    #[test]
    fn typeclass() {
        assert_snapshot!(index_module(indoc!(
            "
        module Test where
        
        class Show a where
          show :: a -> String
        "
        )));
    }
}
