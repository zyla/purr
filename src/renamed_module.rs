use crate::ast::AbsoluteName;
use crate::ast::DeclarationRefConstructors;
use crate::indexed_module::TypeClassDecl;
use crate::indexed_module::TypeDecl;
use fxhash::FxHashMap;
use std::collections::{HashMap, HashSet};

use salsa::DebugWithDb;

use petgraph::{algo::tarjan_scc, prelude::DiGraph};

use crate::{
    ast::{Declaration, DeclarationRefKind, ImportDeclarationKind},
    indexed_module::{IndexedModule, ValueDecl},
    rename::rename_module,
    symbol::Symbol,
    Db, ModuleId, ParsedModule,
};

#[salsa::interned]
pub struct DeclId {
    pub namespace: Namespace,
    pub module: ModuleId,
    pub name: Symbol,
}

impl DeclId {
    fn to_absolute_name(self, db: &dyn Db) -> AbsoluteName {
        AbsoluteName::new(db, self.module(db), self.name(db))
    }
}

#[derive(PartialEq, Eq, Clone, Debug, DebugWithDb, Hash)]
pub enum Namespace {
    Class,
    Type,
    Value,
}

#[derive(PartialEq, Eq, Clone, Debug, DebugWithDb)]
pub struct RenamedModule {
    pub module_id: ModuleId,
    pub imported: Vec<(Option<ModuleId>, DeclId)>,
    pub exported: Vec<DeclId>,
    pub types: FxHashMap<AbsoluteName, TypeDecl>,
    pub values: FxHashMap<AbsoluteName, ValueDecl>,
    pub classes: FxHashMap<AbsoluteName, TypeClassDecl>,
}

#[salsa::tracked]
pub fn renamed_value_decl(db: &dyn Db, id: DeclId) -> ValueDecl {
    assert!(id.namespace(db) == Namespace::Value);
    // FIXME: we shoudn't have to clone here
    let abs_name = id.to_absolute_name(db);
    renamed_module(db, id.module(db))
        .values
        .get(&abs_name)
        .cloned()
        .unwrap_or_else(|| {
            panic!(
                "renamed_value_decl for nonexistent value {:?}",
                abs_name.into_debug_all(db)
            )
        })
}

#[salsa::tracked]
pub fn renamed_module(db: &dyn Db, module_id: ModuleId) -> RenamedModule {
    let mut indexed = crate::indexed_module::indexed_module(db, module_id);
    let mut imported = crate::renamed_module::imported_decls(db, module_id);
    let mut exported = crate::renamed_module::exported_decls(db, module_id);
    let mut diagnositics = vec![];

    let module = crate::parsed_module(db, module_id);

    rename_module(
        db,
        &mut indexed,
        &mut imported,
        &mut exported,
        &mut diagnositics,
    );

    let mut graph = DiGraph::<Declaration, ()>::new();
    let mut node_indices = HashMap::new();

    for declaration in module.ast.declarations.iter() {
        let node_index = graph.add_node(declaration.clone());
        node_indices.insert(declaration.clone(), node_index);
    }

    // TODO: Add edges between nodes based on data structure
    // for declaration in module.ast.declarations.iter() {
    //     use crate::ast::DeclarationKind;
    //     match &***declaration {
    //         // Handle each variant of DeclarationKind
    //         DeclarationKind::Data { constructors, .. } => {
    //             for constructor in constructors {
    //                 if let Some(target_node) = node_indices
    //                     .get(&DeclarationKind::ValueDeclaration(constructor.name.clone()))
    //                 {
    //                     graph.add_edge(node_indices[declaration], *target_node, ());
    //                 }
    //             }
    //         }
    //         DeclarationKind::TypeSynonym { body, .. } => todo!(),
    //     }
    // }

    let _scc = tarjan_scc(&graph);

    RenamedModule {
        module_id,
        imported: imported.clone(),
        exported: exported.clone(),
        values: indexed.values,
        types: indexed.types,
        classes: indexed.classes,
    }
}

struct ExportedDeclExtractor<'a> {
    db: &'a dyn Db,
    module_id: ModuleId,
    exported_decls: Vec<DeclId>,
    imported_decls: Vec<(Option<ModuleId>, DeclId)>,
}

impl<'a> ExportedDeclExtractor<'a> {
    fn extract(&mut self, module: &ParsedModule, indexed: &IndexedModule) {
        let db = self.db;
        match &module.ast.exports {
            Some(decl_ref_kind) => {
                for ref_decl in decl_ref_kind {
                    use DeclarationRefKind::*;

                    match &**ref_decl {
                        Module { name: qualified_as } => {
                            self.exported_decls.extend(
                                self.imported_decls
                                    .iter()
                                    .filter(|(name, _)| {
                                        name.is_some_and(|name| name == *qualified_as)
                                    })
                                    .map(|(_, decl)| *decl),
                            );
                        }
                        x => {
                            self.exported_decls
                                .extend(to_decls_id(db, self.module_id, x));
                        }
                    }
                }
            }
            None => {
                for abs_name in indexed.values.keys() {
                    self.exported_decls.push(DeclId::new(
                        db,
                        Namespace::Value,
                        indexed.module_id,
                        abs_name.name(db),
                    ));
                }

                for abs_name in indexed.types.keys() {
                    self.exported_decls.push(DeclId::new(
                        db,
                        Namespace::Type,
                        indexed.module_id,
                        abs_name.name(db),
                    ));
                }

                for abs_name in indexed.classes.keys() {
                    self.exported_decls.push(DeclId::new(
                        db,
                        Namespace::Class,
                        indexed.module_id,
                        abs_name.name(db),
                    ));
                }
            }
        }
    }
}

#[salsa::tracked]
pub fn exported_decls(db: &dyn Db, module_id: ModuleId) -> Vec<DeclId> {
    let indexed = crate::indexed_module::indexed_module(db, module_id);
    let module = crate::parsed_module(db, module_id);
    let imported = imported_decls(db, module_id);

    let mut extractor = ExportedDeclExtractor {
        db,
        module_id,
        exported_decls: Default::default(),
        imported_decls: imported,
    };

    extractor.extract(&module, &indexed);

    extractor.exported_decls
}

pub fn imported_decls(db: &dyn Db, module_id: ModuleId) -> Vec<(Option<ModuleId>, DeclId)> {
    let module = crate::parsed_module(db, module_id);

    let mut imports: Vec<(Option<ModuleId>, DeclId)> = Vec::new();

    let prim_module_id = ModuleId::new(db, "Prim".into());

    let has_prim = module
        .ast
        .imports
        .iter()
        .any(|i| i.module == prim_module_id);

    if module_id != prim_module_id && !has_prim {
        crate::renamed_module::exported_decls(db, prim_module_id)
            .iter()
            .for_each(|i| imports.push((None, *i)));
    }

    for import in &module.ast.imports {
        use ImportDeclarationKind::*;
        match &import.kind {
            Implicit => {
                crate::renamed_module::exported_decls(db, import.module)
                    .iter()
                    .for_each(|i| imports.push((import.alias, *i)));
            }
            Explicit(decls) => {
                decls
                    .iter()
                    .flat_map(|i| to_decls_id(db, import.module, i))
                    .for_each(|i| imports.push((import.alias, i)));
            }
            Hiding(decls) => {
                let excluded: HashSet<DeclId> = decls
                    .iter()
                    .map(|i| to_decl_id(db, import.module, i))
                    .collect();
                crate::renamed_module::exported_decls(db, import.module)
                    .iter()
                    .filter(|i| !excluded.contains(i))
                    .for_each(|i| imports.push((import.alias, *i)));
            }
        }
    }

    imports
}

fn to_decls_id(db: &dyn Db, module_id: ModuleId, kind: &DeclarationRefKind) -> Vec<DeclId> {
    use DeclarationRefKind::*;

    match kind {
        Type { name, constructors } => {
            let mut decls = vec![DeclId::new(db, Namespace::Type, module_id, *name)];

            if let Some(constructors) = constructors {
                match constructors {
                    DeclarationRefConstructors::Some(constructors) => {
                        for constructor_name in constructors {
                            decls.push(DeclId::new(
                                db,
                                Namespace::Value,
                                module_id,
                                *constructor_name,
                            ));
                        }
                    }
                    DeclarationRefConstructors::All => {}
                }
            }
            decls
        }
        kind => vec![to_decl_id(db, module_id, kind)],
    }
}

fn to_decl_id(db: &dyn Db, module_id: ModuleId, kind: &DeclarationRefKind) -> DeclId {
    use DeclarationRefKind::*;

    match *kind {
        TypeClass { name } => DeclId::new(db, Namespace::Class, module_id, name),
        TypeOp { name } => DeclId::new(db, Namespace::Type, module_id, name),
        Type { name, .. } => DeclId::new(db, Namespace::Type, module_id, name),
        Value { name } => DeclId::new(db, Namespace::Value, module_id, name),
        ValueOp { name } => DeclId::new(db, Namespace::Value, module_id, name),
        TypeInstanceRef { name, .. } => DeclId::new(db, Namespace::Type, module_id, name),
        Module { .. } => panic!("Cannot map module to DeclId"),
    }
}

#[cfg(test)]
mod tests {
    use crate::Diagnostics;

    use super::*;
    use crate::utils::tests::DropSalsaId;
    use indoc::indoc;
    use insta::{self, assert_snapshot};

    const LIB1: &str = indoc!(
        "
        module Lib where

        data Foo = Bar
        "
    );

    const LIB2: &str = indoc!(
        "
        module Lib2 where
         
        x = 1
        y = 2
        "
    );

    fn export_decls(input: &str, deps: Vec<&str>) -> String {
        let db = &mut crate::Database::test_single_file_db(input);
        let module_id = ModuleId::new(db, "Test".into());

        deps.into_iter().zip(1..).for_each(|(deb, i)| {
            db.add_source_file(format!("Lib{}.purs", i).into(), deb.into())
                .unwrap();
        });

        format!(
            "{:#?}",
            (
                exported_decls(db, module_id).into_debug_all(db),
                renamed_module::accumulated::<Diagnostics>(db, module_id)
            )
        )
        .drop_salsa_id()
    }

    fn import_decls(input: &str, deps: Vec<&str>) -> String {
        let db = &mut crate::Database::test_single_file_db(input);
        let module_id = ModuleId::new(db, "Test".into());
        deps.into_iter().zip(1..).for_each(|(deb, i)| {
            db.add_source_file(format!("Lib{}.purs", i).into(), deb.into())
                .unwrap();
        });

        format!(
            "{:#?}",
            (
                imported_decls(db, module_id).into_debug_all(db),
                renamed_module::accumulated::<Diagnostics>(db, module_id)
            )
        )
        .drop_salsa_id()
    }

    #[test]
    fn export_data_decl() {
        assert_snapshot!(export_decls(
            indoc!(
                "
    module Test (Foo) where
    data Foo
    "
            ),
            vec![]
        ))
    }

    #[test]
    fn export_blanket_data_decl() {
        assert_snapshot!(export_decls(
            indoc!(
                "
    module Test where
    data Foo
    "
            ),
            vec![]
        ))
    }

    #[test]
    fn reexport_subset() {
        assert_snapshot!(export_decls(
            indoc!(
                "
    module Test (module X) where
    import Lib2 (x) as X
    "
            ),
            vec![LIB2]
        ))
    }

    #[test]
    fn reexport_all() {
        assert_snapshot!(export_decls(
            indoc!(
                "
    module Test (module X) where
    import Lib2 as X
    "
            ),
            vec![LIB2]
        ))
    }

    #[test]
    fn reexport_export_list() {
        assert_snapshot!(export_decls(
            indoc!(
                "
    module Test (module X) where
    import Lib as X
    "
            ),
            vec![indoc!(
                "
            module Lib (x) where
            x = 1
            y = 2
            "
            )]
        ))
    }

    #[test]
    #[ignore = "Type classes are not yet supported"]
    fn export_class_decl() {
        assert_snapshot!(export_decls(
            indoc!(
                "
    module Test where

    class Foo a where
      foo :: a
    "
            ),
            vec![]
        ))
    }

    #[test]
    fn import_data_decl() {
        assert_snapshot!(import_decls(
            indoc!(
                "
        module Test where
        import Prim ()    
        import Foo (Foo(..))
        "
            ),
            vec![LIB1]
        ))
    }

    #[test]
    fn import_data_qualified_decl() {
        assert_snapshot!(import_decls(
            indoc!(
                "
        module Test where
        import Prim ()    
        import Lib as Lib
        "
            ),
            vec![LIB1]
        ))
    }

    #[test]
    fn import_all_fns() {
        assert_snapshot!(import_decls(
            indoc!(
                "
        module Test where
        import Prim ()    
        import Lib2
        "
            ),
            vec![LIB2]
        ))
    }

    #[test]
    fn import_subset() {
        assert_snapshot!(import_decls(
            indoc!(
                "
        module Test where
        import Prim ()    
        import Lib2 (x)
        "
            ),
            vec![LIB2]
        ))
    }

    #[test]
    fn reexport_qualified() {
        assert_snapshot!(export_decls(
            indoc!(
                "
            module Test
              ( module L
              , x ) where
            
            import Lib as L
            
            x = 1
            "
            ),
            vec![LIB1]
        ))
    }

    #[test]
    fn reexport_qualified_itself() {
        assert_snapshot!(export_decls(
            indoc!(
                "
            module Test (module L, module Test) where
            
            import Lib as L
            
            x = 1
            "
            ),
            vec![LIB1]
        ))
    }

    #[test]
    #[ignore = "how should we handle this?"]
    fn import_self() {
        assert_snapshot!(export_decls(
            indoc!(
                "
            module Test where
            import Test
            x = 1
            "
            ),
            vec![]
        ))
    }
}
