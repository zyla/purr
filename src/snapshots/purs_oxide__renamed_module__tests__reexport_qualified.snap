---
source: src/renamed_module.rs
expression: "export_decls(indoc!(\"\n            module Test\n              ( module L\n              , x ) where\n            \n            import Lib as L\n            \n            x = 1\n            \"),\n    vec![LIB1])"
---
(
    [
        DeclId {
            namespace: Type,
            module: ModuleId {
                name: "Lib",
            },
            name: Symbol {
                text: "Foo",
            },
        },
        DeclId {
            namespace: Value,
            module: ModuleId {
                name: "Test",
            },
            name: Symbol {
                text: "x",
            },
        },
    ],
    [],
)
